### ISSUE TITLE
`interpolate(volume_function_space, surface_function_space)` does not work with DG function space. 

### ISSUE THREAD

**Issue by keiyamamo** [KVSlab/VaSP#100]
**Problem** (This problem is not part of the repository yet, but posted here for the sake of discussion and as a memo)
When interpolating DG function of a volume mesh to a surface mesh, FEniCS naive `interpolate` function does not work properly. This is due to the fact that FEniCS does not know the ownership of the DOF and assign same DOF to the vertex, resulting in almost CG function. The problem becomes worse in parallel since mesh is distributed differently. 

**What we want**
We have WSS computation over the entire domain, but it is only defined on the boundary. Therefore, we want to extract the value at the boundary and save them with boundary mesh to save IO. 

**Temporal solution**
We can save WSS with volume mesh as a temporal solution. File size becomes big, but it shows correct result.

**MWE**
```python
from dolfin import *

# not local projection at the moment
def local_project(f, V):
    u = TrialFunction(V)
    v = TestFunction(V)
    a_proj = inner(u, v)*ds
    b_proj = inner(f, v)*ds
    A = assemble(a_proj, keep_diagonal=True)
    A.ident_zeros()
    b = assemble(b_proj)
    # solver = LocalSolver(A, b)
    # solver.factorize()
    u = Function(V)
    # solver.solve_local_rhs(u)
    solve(A, u.vector(), b)
    # from IPython import embed; embed(); exit(1)
    return u


class STRESS:
    def __init__(self, u, nu, mesh, velocity_degree):
        boundary_ds = Measure("ds", domain=mesh)
        boundary_mesh = BoundaryMesh(mesh, 'exterior')
        self.bmV = VectorFunctionSpace(boundary_mesh, 'DG', velocity_degree -1)
        self.V = VectorFunctionSpace(mesh, 'DG', velocity_degree -1)

        # Compute stress tensor
        sigma = (2 * nu * sym(grad(u)))
        # Compute stress on surface
        n = FacetNormal(mesh)
        F = -(sigma * n)

        # Compute normal and tangential components
        Fn = inner(F, n)  # scalar-valued
        self.Ft = F 
        #- (Fn * n)  # vector-valued
        #from IPython import embed;embed()

    def __call__(self):
        """
        Compute stress for given velocity field u

        Returns:
            Ftv_mb (Function): Shear stress
        """
        self.Ftv = local_project(self.Ft, self.V)
        self.Ftv_bd = interpolate(self.Ftv, self.bmV)

        return self.Ftv, self.Ftv_bd

velocity_degree = 2
mesh = UnitCubeMesh(10, 10, 10)

V = VectorFunctionSpace(mesh, "CG", velocity_degree)
f = Expression(("10*x[0]*x[0]+sin(x[0])", "cos(x[1])", "x[2]"), degree=velocity_degree)
u_2 = interpolate(f, V)
volume_writer = XDMFFile("u2.xdmf")
volume_writer.write_checkpoint(u_2, "u", 0, XDMFFile.Encoding.HDF5, False)
volume_writer.close()


stress = STRESS(u_2, 1, mesh, velocity_degree)
volume_tau, surface_tau = stress()

# Here, volume_tau is correct, but surface_tau is incorrect due to interpolate()
volume_writer = XDMFFile("volume_tau.xdmf")
volume_writer.write_checkpoint(volume_tau, "tau", 0, XDMFFile.Encoding.HDF5, False)
volume_writer.close()
surface_tau_writer = XDMFFile("surface_tau.xdmf")
surface_tau_writer.write_checkpoint(surface_tau, "tau", 0, XDMFFile.Encoding.HDF5, False)
surface_tau_writer.close()

```

**Comment by jorgensd** [comment]
A working version using MeshView and some dark magic
```python
from IPython import embed
from dolfin import *
import numpy as np
mesh = UnitCubeMesh(10, 10, 5)
marker = MeshFunction("size_t", mesh, mesh.topology().dim()-1, 0)
for f in facets(mesh):
    marker[f] = (np.isclose(f.midpoint().x(), 0) or np.isclose(f.midpoint().x(), 1) or
                 np.isclose(f.midpoint().y(), 0) or np.isclose(f.midpoint().y(), 1))
submesh = create_meshview(marker, 1)

expr = Expression("x[0] + x[1]", degree=1)
V1 = FunctionSpace(mesh, "DG", 1)
u = interpolate(expr, V1)

print(
    f"Value of boundary integral is {assemble(u*ds(domain=mesh, subdomain_data=marker, subdomain_id=1))}")


V_sub = FunctionSpace(submesh, "DG", 1)
u_sub = interpolate(expr, V_sub)
print(f"SubMesh interpolated function {assemble(u_sub*dx(domain=submesh))}")

# Create functin on sub mesh
v_sub = Function(V_sub)

# Get mapping from sub mesh cell to parent facet
sub_map = submesh.topology().mapping()[0].cell_map()

tdim = mesh.topology().dim()
sub_dofmap = V_sub.dofmap()

# Get copies of u vector and v vector (to assign afterwards)
u_vec = u.vector().get_local()
sub_copy = v_sub.vector().get_local()
mesh.init(mesh.topology().dim()-1, mesh.topology().dim())
f_to_c = mesh.topology()(mesh.topology().dim()-1, mesh.topology().dim())

dof_coords = V1.tabulate_dof_coordinates()
for i, facet in enumerate(sub_map):
    f = Facet(mesh, facet)
    vertices = [Vertex(mesh, v_) for v_ in f.entities(0)]
    facet_coords = [v_.point().array() for v_ in vertices]
    cells = f_to_c(facet)
    # Get closure dofs on parent facet
    closure_dofs = V1.dofmap().entity_closure_dofs(
        mesh, mesh.topology().dim(), [cells[0]])
    copy_dofs = []
    for dof in closure_dofs:
        for f_coord in facet_coords:
            if np.allclose(dof_coords[dof], f_coord):
                copy_dofs.append(dof)
                break
    sub_dofs = sub_dofmap.cell_dofs(i)
    # Copy data
    sub_copy[sub_dofs] = u_vec[copy_dofs]
v_sub.vector().set_local(sub_copy)


print(assemble(v_sub*dx(domain=submesh)))

with XDMFFile("v_sub.xdmf") as xdmf:
    xdmf.write_checkpoint(v_sub, "v", 0.0, append=False)

with XDMFFile("u.xdmf") as xdmf:
    xdmf.write_checkpoint(u, "u", 0.0, append=False)
```
![image](https://github.com/KVSlab/VaSP/assets/8554300/328d708b-95e8-4e2d-84cc-ecd062c5bf44)
Using meshview for now. If we can find a mapping from BoundaryMesh-cell to parent mesh facet we could replace it.
